src/main.cpp

```diff
@@ -20,6 +20,8 @@ int main(int argc, char** argv)
 
 		ConfigFileParser parser;
 		ConfigFile configFile = parser.parseConfigFile(argv[1]);
+		if (configFile.servers.empty())
+			return 0;
```

Comment:
Valid question. 

When there is the following config:

```nginx
htpp{}
```

And this command is executed (to check if the config file is correct):
```bash
sudo nginx -t
```

Then the return code is ``0``. 

I also tried to stop the nginx server and start it with config file mentioned above. 
The start then is successful and the status of the server is running. In this case the return code is also ``0``. 

But in the way I implemented it, our webserver does **NOT** start and just returns ``0``. Should I change this behavior?

Reply:
Question is with what server nginx is running in that case. If it is a default server, we could replicate this behaviour by loading a default config into the ConfigFile that we have prepared. Can you figure out what kind of server config nginx is using?

Reply:
I am not quite sure if there is a certain server running in the background (even though obviously the server is up), because there is no response to a request. Not even an error code. 

This is the output that I get when making a GET request by updating the page:

<img width="710" alt="image" src="https://github.com/user-attachments/assets/34d8fc2a-300b-4c2a-bdff-2f8b741ffdd3">

And this is the output of the inspection tool:

<img width="889" alt="image" src="https://github.com/user-attachments/assets/79d47c96-7242-4594-9b67-59da4de543e0">

Therefore it does not seem to allow any connections but the server is still running.

Perhaps we could mimic this behaviour by not allowing any connections to the webserver even though it is running.

Reply:
Well, we could simply have the prog run in a loop which does nothing until a signal is sent:

```if (configFile.servers.empty())
{
   while (g_signalStatus == 0)
      continue;
   if (g_signalStatus == SIGQUIT)
		shutdownServer(server);
}
else {
                configFile = createDummyConfig();

		Server server(configFile, epollWrapper, socketPolicy);
		initVirtualServers(server, 10, server.getServerConfigs());
		runServer(server);
}
```

Reply:
Yes, seems like a good approach to solve in this way. 

I implemented it in this way:
```cpp
ConfigFileParser parser;
		ConfigFile configFile = parser.parseConfigFile(argv[1]);
		if (configFile.servers.empty()) {
			while (g_signalStatus == 0)
				continue;
			if (g_signalStatus == SIGQUIT)
				return 0;
		} else {
		configFile = createDummyConfig();

		Server server(configFile, epollWrapper, socketPolicy);
		initVirtualServers(server, 10, server.getServerConfigs());
		runServer(server);
		}
```
**Important fact**
If you try to establish a connection to localhost with port ``8080`` (either using telnet or a browser), it can be the case, that the connection loads infinitely. 
Even though, our webserv does not run any virtual servers while you do that. 

This is the case, because the devcontainer in  VSCode forwards the port 8080 for localhost. If the connection to the devcontainer is closed, then this port gets not used anymore (except if another program, service etc. is using this port). 

You can check which process uses the port ``8080`` with this command:
```bash
lsof -i :8080
```

Reply:
Actually we can leave away the `if (g_signalStatus == SIGQUIT)`. It is only required for a graceful shutdown. As in this case we don't have any servers or Connections running, nothing needs to be shut down gracefully. Even the return 0 is not necessary as we will anyways reach the return 0 at the end of main().

Reply:
I have been thinking about how to test this case and decided to update the main like this:

```cpp

try {
		EpollWrapper epollWrapper(10, -1);
		SocketPolicy socketPolicy;

		ConfigFileParser parser;
		ConfigFile configFile = parser.parseConfigFile(argv[1]);

		Server server(configFile, epollWrapper, socketPolicy);
		initVirtualServers(server, 10, server.getServerConfigs());
		runServer(server);
	} catch (std::exception& e) {
		LOG_ERROR << e.what();
		return 1;
	}
	return 0;
```

In this case we don't have to check because their are no servers within the config file, ergo no virtual servers get created to which you can establish a connection. 

According to this logic the tests would look like this:

```cpp
TEST_F(InvalidConfigFileTests, FileMissesHtppBlock)
{
	ConfigFile configFile = m_configFileParser.parseConfigFile("config_files/missing_http_block.conf");
	EXPECT_EQ(true, configFile.servers.empty());

	EpollWrapper epollWrapper(10, -1);
	SocketPolicy socketPolicy;

	Server server(configFile, epollWrapper, socketPolicy);

	initVirtualServers(server, 10, server.getServerConfigs());
	std::map<int, Socket> virtualServers = server.getVirtualServers();
	EXPECT_EQ(0, virtualServers.size());
}

TEST_F(InvalidConfigFileTests, FileMissesServerBlock)
{
	ConfigFile configFile = m_configFileParser.parseConfigFile("config_files/missing_server_block.conf");
	EXPECT_EQ(true, configFile.servers.empty());

	EpollWrapper epollWrapper(10, -1);
	SocketPolicy socketPolicy;

	Server server(configFile, epollWrapper, socketPolicy);

	initVirtualServers(server, 10, server.getServerConfigs());
	std::map<int, Socket> virtualServers = server.getVirtualServers();
	EXPECT_EQ(0, virtualServers.size());
}
```

Reply:
You mean no servers get created inside the _initVirtualServers()_ and therefore nothing happens in _runServer()_‽ Is that correct?

If so, then yes, we can leave out the whole check for empty servers in configFile.

Reply:
Exactly, that is what I mean. 

I outcommented the following code:

```cpp
// configFile = createDummyConfig();
```

Then executed this command:
```bash
./webserv config_files/missing_server_block.conf 
```
This the output then:
```bash
2024-11-05 10:20:23  [INFO] Initializing virtual servers
2024-11-05 10:20:23  [DEBUG] <bool initVirtualServers(Server &, int, const std::vector<ConfigServer> &)>(src/Server.cpp:527): Check for wildcard servers
2024-11-05 10:20:23  [DEBUG] <bool initVirtualServers(Server &, int, const std::vector<ConfigServer> &)>(src/Server.cpp:543): Add remaining virtual servers
2024-11-05 10:20:23  [INFO] Server started
2024-11-05 10:20:23  [DEBUG] <virtual int EpollWrapper::waitForEvents()>(src/EpollWrapper.cpp:54): Waiting for events
```

Even though it looks like something is set up it's actual not the case because no virtual servers get initialized and run. 

---

Currently our webserver still starts normally, because we create a DummyConfig until everything is parsed. 
But the server behaves correctly if we use the config file which is created by the parser. 

Therefore I would say, that the main I posted in the comment before should stay and the tests make in the current state sure, that our program behaves correctly.

Reply:
Sounds good, let's leave it then